### PR TITLE
neo4j: 3.5.14 -> 4.1.1

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
wanted to use it for an upcoming project. Had difficulty using the 3.5 version, so i bumped it, and redid some of the module

I'm not a big fan of how the state directory is created, but I'm not very knowledgeable with systemd.

The nixosTests also seem to have a high degree of failure with "no space left on device". Not sure if it's hitting RAM or space issues. Most likely RAM issues.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
Also, the upgrade from 3.5 to 4 looks none trivial. Not sure how we should handle this scenario https://neo4j.com/docs/operations-manual/current/upgrade/planning/

**Comment by 0xB10C** [comment]
Still relevant. Having an up-to-date neo4j version (4.X) would be good.

File: pkgs/servers/nosql/neo4j/default.nix [hunk 1]
```diff
@@ -1,39 +1,42 @@
...
+    patchShebangs $out/share/neo4j/bin/neo4j-admin
+    # user will be asked to change password on first login
+    $out/bin/neo4j-admin set-initial-password neo4j
   '';
 
+  passthru.tests.nixos = nixosTests.neo4j;
+
   meta = with stdenv.lib; {
     description = "A highly scalable, robust (fully ACID) native graph database";
     homepage = "http://www.neo4j.org/";
```
Comment (SuperSandro2000) [inline pkgs/servers/nosql/neo4j/default.nix:37]:
```suggestion
    homepage = "https://neo4j.com/";
```

**Comment by erictapen** [comment]
I'd like to continue this in https://github.com/NixOS/nixpkgs/pull/136300. @jonringer are you okay if I close this PR?

**Pull request closed without merging**